components/homme/src/theta-l_kokkos/cxx/LimiterFunctor.hpp

```diff
@@ -0,0 +1,214 @@
+  using TeamPolicyType = Kokkos::TeamPolicy<ExecSpace,Kokkos::LaunchBounds<512,1>,Tag>;
+#else
+  template<typename Tag>
+  using TeamPolicyType = Kokkos::TeamPolicy<ExecSpace,Tag>;
+#endif
+
+  TeamPolicyType<TagDp3dLimiter>   m_policy_dp3d_lim;
+
+  TeamUtils<ExecSpace> m_tu;
+
+  LimiterFunctor(const Elements &elements, const HybridVCoord &hvcoord, const SimulationParams& params)
+      : m_num_elems(elements.num_elems())
+      , m_theta_hydrostatic_mode(params.theta_hydrostatic_mode)
+      , m_hvcoord(hvcoord)
+      , m_state(elements.m_state)
+      , m_geometry(elements.m_geometry)
+      , m_policy_dp3d_lim (Homme::get_default_team_policy<ExecSpace,TagDp3dLimiter>(m_num_elems))
+      , m_tu(m_policy_dp3d_lim)
+      , m_dp3d_thresh(params.dp3d_thresh)
+      , m_vtheta_thresh(params.vtheta_thresh)
+  {
+    m_np1 = -1;
+  }
+
+  int requested_buffer_size () const {
+    // Ask the buffers manager to allocate enough buffers
+    const int nslots = m_tu.get_num_ws_slots();
+
+    int num_scalar_mid_buf = Buffers::num_3d_scalar_mid_buf;
+
+    return num_scalar_mid_buf  *NP*NP*NUM_LEV  *VECTOR_SIZE*nslots ;
+  }
+
+  void init_buffers (const FunctorsBuffersManager& fbm) {
+    Errors::runtime_check(fbm.allocated_size()>=requested_buffer_size(), "Error! Buffers size not sufficient.\n");
+
+    Scalar* mem = reinterpret_cast<Scalar*>(fbm.get_memory());
+    const int nslots = m_tu.get_num_ws_slots();
+
+    // Midpoints scalars
+    m_buffers.buffer1       = decltype(m_buffers.buffer1)(mem,nslots);
+    mem += m_buffers.buffer1.size();
+
+    assert ((reinterpret_cast<Real*>(mem) - fbm.get_memory())==requested_buffer_size());
+  }
+
+
+  void run (const int& tl)
+  {
+    profiling_resume();
+
+    GPTLstart("caar limiter");
+    m_np1 = tl;
+    Kokkos::parallel_for("caar loop dp3d limiter", m_policy_dp3d_lim, *this);
+    Kokkos::fence();
+    GPTLstop("caar limiter");
+
+    profiling_pause();
+  }
+
+
+//NEED TO THROW IN SOME MESSAGE
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const TagDp3dLimiter&, const TeamMember &team) const {
+    KernelVariables kv(team, m_tu);
+
+    // TODO: make this less hard-coded maybe?
+    //constexpr double dp3d_thresh = 1.0;
+    //constexpr double vtheta_thresh = 400; // Kelvin
+
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(kv.team,NP*NP),
+                         [&](const int idx) {
+      const int igp = idx / NP;
+      const int jgp = idx % NP;
+
+      const auto& spheremp = m_geometry.m_spheremp(kv.ie,igp,jgp);
+
+      // Check if the minimum dp3d in this column is blow a certain threshold
+      auto dp = Homme::subview(m_state.m_dp3d,kv.ie,m_np1,igp,jgp);
+      auto& dp0 = m_hvcoord.dp0;
+      auto diff = Homme::subview(m_buffers.buffer1,kv.team_idx,igp,jgp);
+      Kokkos::parallel_for(Kokkos::ThreadVectorRange(kv.team,NUM_LEV),
+                           [&](const int ilev) {
+        diff(ilev) = (dp(ilev) - m_dp3d_thresh*dp0(ilev))*spheremp;
+      });
+
```

Comment:
Technically need a team_barrier here since in principle NUM_LEV /= NUM_PHYSICAL_LEV.

Reply:
thanks, done